lgc/patch/NggPrimShader.cpp

```diff
@@ -5998,6 +6010,70 @@ Value *NggPrimShader::doDppUpdate(Value *oldValue, Value *srcValue, unsigned dpp
+  if (m_hasGs) {
+    // ES-GS
+    auto &inOutUsage = m_pipelineState->getShaderResourceUsage(ShaderStageGeometry)->inOutUsage;
+    assert(inOutUsage.builtInOutputLocMap.find(BuiltInPosition) != inOutUsage.builtInOutputLocMap.end());
+    const unsigned loc = inOutUsage.builtInOutputLocMap[BuiltInPosition];
+    const unsigned rasterStream = inOutUsage.gs.rasterStream;
+
+    const unsigned vertexItemSize = 4 * inOutUsage.gs.outLocCount[rasterStream];
+    auto vertexOffset = m_builder->CreateMul(vertexId, m_builder->getInt32(vertexItemSize));
+    vertexOffset = m_builder->CreateShl(vertexOffset, 2);
+
+    return importGsOutput(VectorType::get(m_builder->getFloatTy(), 4), loc, 0, rasterStream, vertexOffset);
+  } else {
+    // ES-only
+    return readPerThreadDataFromLds(VectorType::get(m_builder->getFloatTy(), 4), vertexId, LdsRegionPosData);
+  }
+}
+
+// =====================================================================================================================
+// Fetches the aggregated sign mask of cull distances for the specified vertex ID.
+//
+// @param vertexId : Vertex thread ID in sub-group.
+Value *NggPrimShader::fetchCullDistanceSignMask(Value *vertexId) {
+  assert(m_nggControl->enableCullDistanceCulling);
+
+  if (m_hasGs) {
+    // ES-GS
+    auto &inOutUsage = m_pipelineState->getShaderResourceUsage(ShaderStageGeometry)->inOutUsage;
+    assert(inOutUsage.builtInOutputLocMap.find(BuiltInCullDistance) != inOutUsage.builtInOutputLocMap.end());
+    const unsigned loc = inOutUsage.builtInOutputLocMap[BuiltInCullDistance];
+    const unsigned rasterStream = inOutUsage.gs.rasterStream;
+
+    const unsigned vertexItemSize = 4 * inOutUsage.gs.outLocCount[rasterStream];
+    auto vertexOffset = m_builder->CreateMul(vertexId, m_builder->getInt32(vertexItemSize));
+    vertexOffset = m_builder->CreateShl(vertexOffset, 2);
+
+    auto &builtInUsage = m_pipelineState->getShaderResourceUsage(ShaderStageGeometry)->builtInUsage.gs;
+    auto cullDistances = importGsOutput(ArrayType::get(m_builder->getFloatTy(), builtInUsage.cullDistance), loc, 0,
+                                        rasterStream, vertexOffset);
+
+    // Calculate the sign mask for all cull distances
+    Value *signMask = m_builder->getInt32(0);
+    for (unsigned i = 0; i < builtInUsage.cullDistance; ++i) {
+      auto cullDistance = m_builder->CreateExtractValue(cullDistances, i);
+      cullDistance = m_builder->CreateBitCast(cullDistance, m_builder->getInt32Ty());
+
+      Value *signBit = m_builder->CreateIntrinsic(Intrinsic::amdgcn_ubfe, m_builder->getInt32Ty(),
+                                                  {cullDistance, m_builder->getInt32(31), m_builder->getInt32(1)});
+      signBit = m_builder->CreateShl(signBit, i);
+      signMask = m_builder->CreateOr(signMask, signBit);
+
+      return signMask;
+    }
+  } else {
+    // ES-only
+    return readPerThreadDataFromLds(m_builder->getInt32Ty(), vertexId, LdsRegionCullDistance);
+  }
```

Comment:
Same as above.
In addition, it seems like the for loop does only 1 iteration?

Reply:
The for loop depends on cull distance array size. Cull distance is often declared as `float gl_CullDistance[]`. The processing is to handle each cull distance values and check its sign. Then, the signs are changed to a bit mask for future use. The bit layout is like this:

`Sign of CullDistance[7] |  Sign of CullDistance[6] | ... | Sign of CullDistance[0]`

Reply:
Taking a step back: just by looking at the code, the last line of the for loop is a `return`, and there is no `continue` or `break`. This code looks suspicious because it will always return at the end of the first iteration. I don't get it. I'd guess that the `return` was supposed to be outside of the loop, since each iteration updates a variable defined just outside of the loop.

Reply:
Oh. Thanks a lot for catching this. It is my coding error.